### PR TITLE
Use PathBuf instead of String where applicable

### PR DESCRIPTION
**Pull request by oli-obk**
r? @jseyfried

### CODE REVIEW

File: src/librustc_driver/driver.rs [hunk 1]
```diff
@@ -306,17 +307,10 @@ fn keep_hygiene_data(sess: &Session) -> bool {
...
-/// The name used for source code that doesn't originate in a file
-/// (e.g. source from stdin or a string)
-pub fn anon_src() -> String {
-    "<anon>".to_string()
-}
-
-pub fn source_name(input: &Input) -> String {
+pub fn source_name(input: &Input) -> FileName {
     match *input {
         // FIXME (#9639): This needs to handle non-utf8 paths
```
Comment (kennytm) [inline src/librustc_driver/driver.rs]: Is this FIXME still relevant in this place?

Comment (oli-obk) [reply src/librustc_driver/driver.rs]: I think it's fixed with this PR, but there's no repro example on the issue

**Comment by oli-obk** [comment]
Travis failure is legit

**Comment by oli-obk** [comment]
Travis likes it now

**Comment by oli-obk** [comment]
rebased

**Comment by jseyfried** [comment]
Excellent, looks great! Will do detailed review tomorrow.

File: src/libsyntax/parse/parser.rs [hunk 2]
```diff
@@ -528,7 +528,11 @@ impl<'a> Parser<'a> {
         if let Some(directory) = directory {
             parser.directory = directory;
         } else if parser.span != syntax_pos::DUMMY_SP {
-            parser.directory.path = sess.codemap().span_to_unmapped_path(parser.span);
+            // FIXME: this conversion from FileName to PathBuf is weird
+            parser.directory.path = sess.codemap()
+                                        .span_to_unmapped_path(parser.span)
+                                        .to_string()
+                                        .into();
```
Comment (jseyfried) [inline src/libsyntax/parse/parser.rs]: Why do we need `PathBuf` -> `String` -> `PathBuf` here?

Comment (oli-obk) [reply src/libsyntax/parse/parser.rs]: lol. I have no idea what I did there. That was probably some intermediate step to get stuff to compile.

Comment (oli-obk) [reply src/libsyntax/parse/parser.rs]: Ah no, I remember why. This is `FileName -> String -> PathBuf`. I should probably change the `path` field to be a `FileName`

Comment (oli-obk) [reply src/libsyntax/parse/parser.rs]: And fixed

**Comment by kennytm** [comment]
There are two `run-pass-fulldeps` tests failed with "position 0 does not resolve to a source location".

<details>

```
[01:01:07] ---- [run-pass] run-pass-fulldeps/issue-35829.rs stdout ----
[01:01:07] 	
[01:01:07] error: test run failed!
[01:01:07] status: exit code: 101
[01:01:07] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass-fulldeps/issue-35829.stage2-x86_64-unknown-linux-gnu"
[01:01:07] stdout:
[01:01:07] ------------------------------------------
[01:01:07] 
[01:01:07] ------------------------------------------
[01:01:07] stderr:
[01:01:07] ------------------------------------------
[01:01:07] thread 'main' panicked at 'Some tests failed', /checkout/src/tools/compiletest/src/main.rs:331:21
[01:01:07] thread 'main' panicked at 'position 0 does not resolve to a source location', /checkout/src/libsyntax/codemap.rs:632:8
[01:01:07] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:01:07] 
[01:01:07] ------------------------------------------
[01:01:07] 
[01:01:07] thread '[run-pass] run-pass-fulldeps/issue-35829.rs' panicked at 'explicit panic', /checkout/src/tools/compiletest/src/runtest.rs:2570:8
[01:01:07] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:01:07] 
[01:01:07] ---- [run-pass] run-pass-fulldeps/qquote.rs stdout ----
[01:01:07] 	
[01:01:07] error: test run failed!
[01:01:07] status: exit code: 101
[01:01:07] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass-fulldeps/qquote.stage2-x86_64-unknown-linux-gnu"
[01:01:07] stdout:
[01:01:07] ------------------------------------------
[01:01:07] 
[01:01:07] ------------------------------------------
[01:01:07] stderr:
[01:01:07] ------------------------------------------
[01:01:07] thread 'main' panicked at 'position 0 does not resolve to a source location', /checkout/src/libsyntax/codemap.rs:632:8
[01:01:07] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:01:07] 
[01:01:07] ------------------------------------------
[01:01:07] 
[01:01:07] thread '[run-pass] run-pass-fulldeps/qquote.rs' panicked at 'explicit panic', /checkout/src/tools/compiletest/src/runtest.rs:2570:8
[01:01:07] 
[01:01:07] 
[01:01:07] failures:
[01:01:07]     [run-pass] run-pass-fulldeps/issue-35829.rs
[01:01:07]     [run-pass] run-pass-fulldeps/qquote.rs
[01:01:07] 
[01:01:07] test result: FAILED. 84 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out
```

</details>

**Comment by oli-obk** [comment]
I assumed (wrongly) that `span_to_unmapped_path` handled `DUMMY_SP`, reintroduced the check that was there berfore

**Comment by jseyfried** [comment]
Makes sense! @bors r+

**Comment by oli-obk** [comment]
Legit. Aux builder pretty-fulldeps

**Comment by oli-obk** [comment]
That was a weird one. `./x.py test` never catches it. `./x.py test src/run-pass-fulldeps/pretty` doesn't catch it if `./x.py test` ran through successfully before.

**Comment by kennytm** [comment]
@oli-obk That sounds like a rustbuild bug...

**Comment by oli-obk** [comment]
@kennytm I'm not sure, since it's explicitly done on the aux builder, this might be intentional

**Comment by kennytm** [comment]
@oli-obk The `./x.py test` never catches it is intentional, but `./x.py test src/run-pass-fulldeps/pretty` doesn't catch it is certainly not. But I'm assuming you're running it in `--stage 2` since some tests can't be run in earlier stages.

**Comment by kennytm** [comment]
@bors r=jseyfried

Travis is green, and new commit looks reasonable, let's try again.

**Comment by oli-obk** [comment]
Seems like this breaks rustfmt. I'll open a PR against them

**Comment by oli-obk** [comment]
DO NOT MERGE

until https://github.com/rust-lang-nursery/rustfmt/pull/2251 and https://github.com/rust-lang-nursery/rls/pull/620 have been put on a branch

**Comment by oli-obk** [comment]
Everything is on a branch and worked locally a few hours ago. Let's see what travis says

**Comment by oli-obk** [comment]
Travis is happy. The only thing that changed from the last review is the submodules.

**Comment by kennytm** [comment]
@bors retry 

AppVeyor can't find rust-lang-nursery/rls@60506d9e21df72981f224a170f2e3c3f0bcee221. If the error persist, I suggest you just revert the submodule changes, break the tools, merge the tool PRs, and then submit another Rust PR to unbreak the tool.

**Comment by oli-obk** [comment]
yea no need to retry. I probably setup the wrong commit with all the branch/PR changes

**Comment by oli-obk** [comment]
broke rustfmt and rls and removed submodule changes

cc @nrc

File: src/Cargo.lock [hunk 3]
```diff
@@ -1531,7 +1531,7 @@ dependencies = [
  "rls-rustc 0.1.1 (registry+https://github.com/rust-lang/crates.io-index)",
  "rls-span 0.4.0 (registry+https://github.com/rust-lang/crates.io-index)",
  "rls-vfs 0.4.4 (registry+https://github.com/rust-lang/crates.io-index)",
- "rustfmt-nightly 0.2.17",
+ "rustfmt-nightly 0.3.0",
```
Comment (kennytm) [inline src/Cargo.lock]: Please also revert `Cargo.lock`

**Comment by oli-obk** [comment]
done

**Comment by nrc** [comment]
@oli-obk RLS and Rustfmt have both been fixed (using your commits), could you send a PR to update them in the Rust repo and unbreak them for testing (if you only touch the submodules and Cargo.lock, then pre-emptive r=me, p=1).

**Comment by oli-obk** [comment]
on it

**Pull request merged**